File: snow_pkg/src/simplace/TempMin.java
import  java.io.*;
import  java.util.*;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import javafx.util.*;
import java.time.LocalDateTime;
import java.time.LocalDateTime;
import net.simplace.sim.model.FWSimComponent;
import net.simplace.sim.util.FWSimVarMap;
import net.simplace.sim.util.FWSimVariable;
import net.simplace.sim.util.FWSimVariable.CONTENT_TYPE;
import net.simplace.sim.util.FWSimVariable.DATA_TYPE;
import org.jdom.Element;


public class Tempmin extends FWSimComponent
{
    private FWSimVariable<Double> prof;
    private FWSimVariable<Double> tminseuil;
    private FWSimVariable<Double> tmaxseuil;
    private FWSimVariable<Double> tmin;
    private FWSimVariable<Double> Sdepth_cm;
    private FWSimVariable<Double> tminrec;

    public Tempmin(String aName, HashMap<String, FWSimVariable<?>> aFieldMap, HashMap<String, String> aInputMap, Element aSimComponentElement, FWSimVarMap aVarMap, int aOrderNumber)
    {
        super(aName, aFieldMap, aInputMap, aSimComponentElement, aVarMap, aOrderNumber);
    }

    public Tempmin(){
        super();
    }
    @Override
    protected void process()
    {
        double tSdepth_cm = Sdepth_cm.getValue();
        double tprof = prof.getValue();
        double ttmin = tmin.getValue();
        double ttminseuil = tminseuil.getValue();
        double ttmaxseuil = tmaxseuil.getValue();
        double ttminrec;
        ttminrec = ttmin;
        if (tSdepth_cm > tprof)
        {
            if (ttmin < ttminseuil)
            {
                ttminrec = ttminseuil;
            }
            else
            {
                if (ttmin > ttmaxseuil)
                {
                    ttminrec = ttmaxseuil;
                }
            }
        }
        else
        {
            if (tSdepth_cm > 0.0d)
            {
                ttminrec = ttminseuil - ((1 - (tSdepth_cm / tprof)) * (Math.abs(ttmin) + ttminseuil));
            }
        }
        tminrec.setValue(ttminrec, this);
    }

    @Override
    protected void init()
    {
    }

    @Override
    protected FWSimComponent clone(FWSimVarMap aVarMap)
    {
        return new Tempmin(iName, iFieldMap, iInputMap, iSimComponentElement, aVarMap, iOrderNumber);
    }

    @Override
    public HashMap<String, FWSimVariable<?>> createVariables()
    {
        addVariable(FWSimVariable.createSimVariable("cprof", "snow cover threshold for snow insulation ", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", 0.0, 1000, 0.0, this));
        addVariable(FWSimVariable.createSimVariable("ctminseuil", "minimum temperature when snow cover is higher than prof", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", 0.0, 5000.0, 0.0, this));
        addVariable(FWSimVariable.createSimVariable("ctmaxseuil", "maximum temperature when snow cover is higher than prof", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"http://www.wurvoc.org/vocabularies/om-1.8/", null, null, 0.0, this));
        addVariable(FWSimVariable.createSimVariable("itmin", "current minimum air temperature", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"http://www.wurvoc.org/vocabularies/om-1.8/", 0.0, 100.0, 0.0, this));
        addVariable(FWSimVariable.createSimVariable("sSdepth_cm", "snow depth", DATA_TYPE.DOUBLE, CONTENT_TYPE.state,"http://www.wurvoc.org/vocabularies/om-1.8/", 0.0, 500.0, 0.0, this));
        addVariable(FWSimVariable.createSimVariable("stminrec", "recalculated minimum temperature", DATA_TYPE.DOUBLE, CONTENT_TYPE.state,"http://www.wurvoc.org/vocabularies/om-1.8/", 0.0, 500.0, null, this));

        return iFieldMap;
    }
}

File: snow_pkg/src/openalea/TempMin.py
# coding: utf8
from copy import copy

import numpy
from math import *

def model_TempMin(Sdepth_cm = 0.0,
         prof = 0.0,
         tmin = 0.0,
         tminseuil = 0.0,
         tmaxseuil = 0.0):
    """
     - Name: TempMin -Version: 1.0, -Time step: 1
     - Description:
                 * Title: Minimum temperature  calculation
                 * Author: STICS
                 * Reference: -
                 * Institution: INRA
                 * Abstract: -
     - inputs:
                 * name: Sdepth_cm
                               ** min : 0.0
                               ** default : 0.0
                               ** max : 500.0
                               ** uri : 
                               ** variablecategory : state
                               ** datatype : DOUBLE
                               ** inputtype : variable
                               ** unit : cm
                               ** description : snow depth
                 * name: prof
                               ** parametercategory : constant
                               ** min : 0.0
                               ** datatype : DOUBLE
                               ** max : 1000
                               ** uri : 
                               ** default : 0.0
                               ** inputtype : parameter
                               ** unit : cm
                               ** description : snow cover threshold for snow insulation 
                 * name: tmin
                               ** min : 0.0
                               ** default : 0.0
                               ** max : 100.0
                               ** uri : 
                               ** variablecategory : auxiliary
                               ** datatype : DOUBLE
                               ** inputtype : variable
                               ** unit : degC
                               ** description : current minimum air temperature
                 * name: tminseuil
                               ** parametercategory : constant
                               ** min : 0.0
                               ** datatype : DOUBLE
                               ** max : 5000.0
                               ** uri : 
                               ** default : 0.0
                               ** inputtype : parameter
                               ** unit : degC
                               ** description : minimum temperature when snow cover is higher than prof
                 * name: tmaxseuil
                               ** parametercategory : constant
                               ** min : 
                               ** datatype : DOUBLE
                               ** max : 
                               ** uri : 
                               ** default : 0.0
                               ** inputtype : parameter
                               ** unit : degC
                               ** description : maximum temperature when snow cover is higher than prof
     - outputs:
                 * name: tminrec
                               ** min : 0.0
                               ** variablecategory : state
                               ** max : 500.0
                               ** uri : 
                               ** datatype : DOUBLE
                               ** unit : degC
                               ** description : recalculated minimum temperature
    """

    tminrec = tmin
    if Sdepth_cm > prof:
        if tmin < tminseuil:
            tminrec = tminseuil
        else:
            if tmin > tmaxseuil:
                tminrec = tmaxseuil
    else:
        if Sdepth_cm > 0.0:
            tminrec = tminseuil - ((1 - (Sdepth_cm / prof)) * (abs(tmin) + tminseuil))
    return tminrec